torch/csrc/dynamo/compiled_autograd_py.cpp

```diff
@@ -0,0 +1,526 @@
+#include <torch/csrc/dynamo/compiled_autograd_py.h>
+
+#include <torch/csrc/autograd/compiled_autograd.h>
+#include <torch/csrc/autograd/engine.h>
+#include <torch/csrc/jit/python/pybind_utils.h>
+#include <torch/csrc/python_headers.h>
+#include <torch/csrc/utils/pythoncapi_compat.h>
+#include <iostream>
+#include <vector>
+
+/*
+[Note: Compiled Autograd]
+
+Compiled autograd replaces the standard autograd engine by converting
+the autograd graph to an FX graph that can be torch.compiled. It caches
+this conversion using a shadow graph. We compare the new graph to the
+shadow graph by walking the two graphs simultaneously and computing a
+CacheKey for each original node to find the next edge in the shadow graph.
+Two different graphs might have a shared common prefix in the shadow
+graph, but then diverge at the first difference. Tensors, SavedVariables,
+and SymInt found stored on the nodes in the autograd graph are lifted to
+become inputs to the graph. All other properties (ints, floats, types,
+etc.) are specialized using the CacheKey and will result in landing on
+a different cache node in the shadow graph if some property differs.
+
+To interact with the (hundreds) of different autograd::Node types,
+we use a visitor pattern that walks each Node structure recursively.
+
+- The first pass, compiled_args/collect, extracts all the inputs to the
+graph and builds a CacheKey for us to specialize on.  On a cache hit,
+we stop here and this is the only pass.
+
+- On a cache miss, a second pass kicks in to extract the FX grap using
+apply_with_saved, which uses another visitor pattern.  The before()
+visitor swaps out all the Tensors, SavedVariables, and SymInt for
+fake/symbolic versions to allow tracing.  We then run the standard apply()
+method, and after() restores things to how we found them.
+
+When we see tensor hooks, we record them directly in the output graph
+without tracing into them.  We do this to avoid executing unsafe code
+at trace time.
```

Comment:
Notes:
* Comment about the well behavedness requirements on hooks, IIRC we need to assume they don't do too strange things to the output tensors so we can continue propagating them through with actual fake tensors.
* Comment about the fact that autograd backwards graph are "well behaved"; e.g., we could imagine that the backward graph should just be a bunch of C++ autograd backward functions strung together, but these are always safe to trace so we eagerly trace them out to get a more manageable graph